p2p/lightchain.py

```diff
@@ -315,14 +320,46 @@ def callback(r):
         return rlp.decode(rlp_account, sedes=Account)
 
     @alru_cache(maxsize=1024, cache_exceptions=False)
-    async def get_contract_code(self, block_hash: Hash32, key: bytes) -> bytes:
+    @service_timeout(COMPLETION_TIMEOUT)
+    async def get_contract_code(self, block_hash: Hash32, address: Address) -> bytes:
+        """
+        :param block_hash: find code as of the block with block_hash
+        :param address: which contract to look up
+
+        :return: bytecode of the contract, ``b''`` if no code is set
+        """
         peer = cast(LESPeer, self.peer_pool.highest_td_peer)
+
+        # get account for later verification, and
+        # to confirm that our highest total difficulty peer has the info
+        try:
+            account = await self.get_account(block_hash, address)
+        except HeaderNotFound as exc:
+            raise NoEligiblePeers("Our best peer does not have header %s" % block_hash) from exc
+
+        # request contract code
         request_id = gen_request_id()
-        peer.sub_proto.send_get_contract_code(block_hash, key, request_id)
+        peer.sub_proto.send_get_contract_code(block_hash, keccak(address), request_id)
         reply = await self._wait_for_reply(request_id)
+
         if not reply['codes']:
-            return b''
-        return reply['codes'][0]
+            bytecode = b''
+        else:
+            bytecode = reply['codes'][0]
+
+        # validate bytecode against a proven account
+        if account.code_hash == keccak(bytecode):
+            return bytecode
+        elif bytecode == b'':
+            # TODO disambiguate failure types here, and raise the appropriate exception
+            # An (incorrectly) empty bytecode might indicate a bad-acting peer, or it might not
+            raise NoEligiblePeers("Our best peer incorrectly responded with an empty code value")
+        else:
+            # a bad-acting peer sent an invalid non-empty bytecode
+            # disconnect from the peer
+            await self.disconnect_peer(peer, DisconnectReason.subprotocol_error)
+            # try again with another peer
+            return await self.get_contract_code(block_hash, address)
```

Comment:
I still feel reasonably strongly that this pattern is going to bite us.  Recursive function calls are generally in a similar bucket as `global` and `eval` and this doesn't strike me as one of the situations where it's necessary.  It doesn't seem that big a deal to add proper exception handling and orchestration to the call sites for this function.

Reply:
>Recursive function calls are generally in a similar bucket as global and eval

Yep, strong disagree here, too ;) I find recursion to be quite elegant tbh and it is in fact, the [go to approach in the functional programming camp](https://softwareengineering.stackexchange.com/questions/149167/are-functional-languages-better-at-recursion/149189#149189)

That said, recursion can be dangerous especially in languages that don't do [tail call optimizations](https://stackoverflow.com/questions/310974/what-is-tail-call-optimization)

Anyway, I'm not saying I'm in favor for one solution over the other here. Just wanted to throw in my opinion on recursion.

Reply:
I'll back down on putting `globals/eval` and recursiveness in the same bucket.  The point I was trying to make here was that the **current** implementation allows for an infinite amount of recursion and I very much dislike that.

To @carver point about forgetting to catch the exception.  Fair, and very valid.

At minimum, I'd like to go with the exception based approach for now, and document it's deficiencies and search for a 3rd option that we all like and that doesn't suffer from either of these problems.  I prefer something we might forget over possibly infinitely recursive async calls.

And one last argument against the current approach (which I think I already touched on).  It mixes two types of logic that I don't think belong together.  The *business* part of fetching and validating, and the orchestration part of managing re-issuing the request.

And one last suggestion for a temporary 3rd option.  don't make it recursive.  Just put the whole thing in a loop and only allow a max of X retries and then fail hard if they all fail.